lightning/src/routing/router.rs

```diff
@@ -1521,12 +1529,91 @@ where L::Target: Logger {
+	let network_nodes = network_graph.nodes();
+
+	for path in route.paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..8]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length {
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len())
+						.and_then(|index| cur_node.channels.get(index))
+						.and_then(|id| network_channels.get(id)) {
+							if random_channel.node_one == cur_node_id {
+								cur_node_id = random_channel.node_two;
+								if let Some(dir_info) = &random_channel.one_to_two {
+									shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+										.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+								}
+							} else {
+								cur_node_id = random_channel.node_one;
+								if let Some(dir_info) = &random_channel.two_to_one {
+									shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+										.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+								}
+							};
+						}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40,
+			// which is the most prevalent cltv_expiry_delta currently used in the network
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the total offset to reduce the worst-case locked liquidity timevalue
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 3*144;
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET);
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).max().unwrap_or(0))
```

Comment:
Shouldn't we sum the `cltv_expiry_delta`s in each hop instead of finding the maxiumum one?

Reply:
Thanks for bringing this up. The thought here was that the `max` is even more restrictive of the offset. However, through this I realized I still also used the `max` instead of the `sum` for calculating the `max_path_offset`, which was a remnant from when the offset used to be added to all hops along the route, not only the final one. Fixed now with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/22b4eeb25042813ba905f4d7ea68216624f62aeb.